File: sailrobot/src/controller.cpp
#include "controller.hpp"
#include <utilities.hpp>
#include <sstream>
#include <string>

using namespace Sailboat;

Controller::Controller(std::string name, int looprate, int controller) : name(name), looprate(looprate), controller(controller), rudderAngle(0), sailAngle(0), xbeeMode(0), isSetup(false), beginLoop(false){

}

void Controller::init(int argc, char **argv){
	ros::init(argc, argv, name);

	n = new ros::NodeHandle("~");
	loop_rate = new ros::Rate(looprate);


	gpsSub = n->subscribe("/sailboat/GPS/fix", 100, &Controller::gpsCallback, this);
	imuSub = n->subscribe("/sailboat/IMU", 100, &Controller::imuCallback, this);
	windSub = n->subscribe("/sailboat/wind", 100, &Controller::windCallback, this);
	sailSub = n->subscribe("/sailboat/sail", 100, &Controller::sailCallback, this);
	rudderSub = n->subscribe("/sailboat/rudder", 100, &Controller::rudderCallback, this);
	rudder2Sub = n->subscribe("/sailboat/rudder2", 100, &Controller::rudder2Callback, this);
	velSub = n->subscribe("/sailboat/IMU_Dv", 100, &Controller::velCallback, this);
	xbeeSub = n->subscribe("xbee_send_mode", 100, &Controller::xbeeCallback, this);
	xbeeCSub = n->subscribe("xbee_send_rudder_sail", 100, &Controller::xbeeRudSailCallback, this);

	odomMsg = n->advertise<nav_msgs::Odometry>("/sailboat/odom", 100);
	pubCmd = n->advertise<geometry_msgs::Twist>("/sailboat/sailboat_cmd", 100);
	pubMsg = n->advertise<std_msgs::String>("/sailboat/sailboat_msg", 10);
	pubLog = n->advertise<std_msgs::String>("/sailboat/pc_log", 10);
	pubMarkerA = n->advertise<geometry_msgs::Point>("/control_send_A", 10);
	pubMarkerB = n->advertise<geometry_msgs::Point>("/control_send_B", 10);

	std::string configPath = "config/config.txt";
	if(n->hasParam("config"))
		n->getParam("config",configPath);
	Utility::Instance().config = Utility::ReadConfig(configPath);

	ros::Duration(0.1).sleep();

	if(!loopUnpublished()){
		timer = n->createTimer(ros::Duration(30.0), &Controller::wakeup, this);
		publishMSG("C" + std::to_string(controller));
	}
}

void Controller::wakeup(const ros::TimerEvent& event){
	publishMSG("C" + std::to_string(controller));
}

void Controller::loop(){
	if(isSetup && !beginLoop){
		setup(n);
		beginLoop = true;
	}

	if(beginLoop){
		if(!loopUnpublished())
			controlPublished();
		publishOdom();

	}
	ros::spinOnce();
	loop_rate->sleep();
}

void Controller::controlPublished(){
	if(xbeeMode != 1)
		publishCMD(control());
	else
		publishCMD(xbeeControl);
}


void Controller::publishOdom(){
	nav_msgs::Odometry odom_msg;
	odom_msg.pose.pose.position.x = gpsMsg.latitude;
	odom_msg.pose.pose.position.y = gpsMsg.longitude;
	odom_msg.pose.pose.position.z = gpsMsg.altitude;

	odom_msg.pose.pose.orientation = imuMsg.orientation;

	odom_msg.twist.twist = velMsg;

	odomMsg.publish(odom_msg);
}

void Controller::publishCMD(geometry_msgs::Twist msg){
	pubCmd.publish(msg);
}

void Controller::publishMSG(std_msgs::String msg){
	pubMsg.publish(msg);
}

void Controller::publishMSG(std::string msg){
	std_msgs::String msgD;
	msgD.data = msg.c_str();
	pubMsg.publish(msgD);
}

void Controller::publishLOG(std_msgs::String msg){
	pubLog.publish(msg);
}

void Controller::publishLOG(std::string msg){
	std_msgs::String msgD;
	msgD.data = msg.c_str();
	pubLog.publish(msgD);
}

void Controller::publishMarkerA(float latitude, float longitude){
	geometry_msgs::Point point;
	point.x = latitude;
	point.y = longitude;
	pubMarkerA.publish(point);
}

void Controller::publishMarkerB(float latitude, float longitude){
	geometry_msgs::Point point;
	point.x = latitude;
	point.y = longitude;
	pubMarkerB.publish(point);
}

void Controller::gps(const gps_common::GPSFix::ConstPtr& msg){
	gpsMsg = *msg;
	isSetup = true;
}

void Controller::imu(const sensor_msgs::Imu::ConstPtr& msg){
	imuMsg = *msg;
}

void Controller::wind(const geometry_msgs::Pose2D::ConstPtr& msg){
	windMsg = *msg;
	/*windMsg.x = msg->x;
	  windMsg.y = msg->y;
	  windAvg.push_back(msg->theta);
	  if(windAvg.size() > 50)
	  windAvg.erase(windAvg.begin());
	  float ws = 0;
	  float wc = 0;
	  float w = 0;
	  for(int i = 0; i < windAvg.size(); ++i){
	  ws += sin(windAvg[i]);
	  wc += cos(windAvg[i]);
	  }
	  w = atan2(ws,wc);
	  windMsg.theta = w;*/
}

void Controller::sail(const std_msgs::Float32::ConstPtr& msg){
	sailAngle = msg->data;
}

void Controller::rudder(const std_msgs::Float32::ConstPtr& msg){
	rudderAngle = msg->data;
}

void Controller::rudder2(const std_msgs::Float32::ConstPtr& msg){
	rudder2Angle = msg->data;
}

void Controller::mode(const std_msgs::UInt32::ConstPtr& msg){
	modeControl = msg->data;
}

void Controller::vel(const geometry_msgs::Twist::ConstPtr& msg){
	velMsg = *msg;
}

void Controller::xbeeCallback(const std_msgs::Float32::ConstPtr& msg){
	xbeeMode = msg->data;
}

void Controller::xbeeRudSailCallback(const geometry_msgs::Twist::ConstPtr& msg){
	xbeeControl = *msg;
}

File: sailrobot/scripts/python_class/controller.py
#!/usr/bin/env python
import rospy
from geometry_msgs.msg import Twist
from std_msgs.msg import String
from gps_common.msg import GPSFix
from sensor_msgs.msg import Imu
from geometry_msgs.msg import Pose2D
from std_msgs.msg import Float32
from nav_msgs.msg import Odometry

import math
import numpy
from enum import Enum
import time
import utilities

class MODE:
	STANDBY, RUDDER_SAIL, RETURN_HOME, HEADING, RC, SAIL_CAP, RUDDER = range(7)

class Controller:
	name = ''
	looprate = 10
	controller = 0

 	windMsg = Pose2D()
 	gpsMsg = GPSFix()
 	imuMsg = Imu()
 	velMsg = Twist()
    sailAngle = 0.0
    rudderAngle = 0.0
    rudder2Angle = 0.0
	windArray = []

	def wakeup(self,event):
        self.publishMSG("C" + str(self.controller))
    def loop(self):
        control = self.control()
        if(control):
            self.publishCMD(control)
        self.publishOdom()

        self.rate.sleep()

    def gps(self,data):
        self.gpsMsg = data

    def imu(self,data):
        self.imuMsg = data

    def wind(self,data):
        self.windMsg = data

#            self.windArray.append(data.theta)
#	    if(len(self.windArray) > 25):
#		del self.windArray[0]
#	    ws = numpy.sum(numpy.sin(self.windArray))
#	    wc = numpy.sum(numpy.cos(self.windArray))
#	    w = math.atan2(ws,wc)
#	    self.windMsg.theta = w
#	    self.windMsg.x = data.x
#	    self.windMsg.y = data.y

    def sail(self,data):
        self.sailAngle = data.data

    def rudder(self,data):
        self.rudderAngle = data.data

    def rudder2(self,data):
        self.rudder2Angle = data.data

    def vel(self,data):
        self.velMsg = data

    def publishOdom(self):
        odom_msg = Odometry();
        odom_msg.pose.pose.position.x = self.gpsMsg.latitude;
        odom_msg.pose.pose.position.y = self.gpsMsg.longitude;
        odom_msg.pose.pose.position.z = self.gpsMsg.altitude;

		odom_msg.pose.pose.orientation = self.imuMsg.orientation;
		odom_msg.twist.twist = self.velMsg;
	    self.odomMsg.publish(odom_msg);

    def publishCMD(self, cmd):
        self.pubCmd.publish(cmd)

    def publishMSG(self, msg):
        self.pubMsg.publish(msg)

    def publishLOG(self, msg):
        self.pubLog.publish(msg)

    def __init__(self, name, looprate, mode):
    	self.name = name
    	self.looprate = looprate
    	self.controller = mode

    	rospy.init_node(self.name,anonymous=True)
    	self.rate = rospy.Rate(self.looprate)

    	rospy.Timer(rospy.Duration(30.0), self.wakeup)

    	self.pubCmd = rospy.Publisher('/sailboat/sailboat_cmd', Twist, queue_size=100)
    	self.pubMsg = rospy.Publisher('/sailboat/sailboat_msg', String, queue_size=10)
    	self.pubLog = rospy.Publisher('/sailboat/pc_log', String, queue_size=10)
    	self.odomMsg = rospy.Publisher('/sailboat/odom', Odometry, queue_size=100)
    	self.gpsSub = rospy.Subscriber('/sailboat/GPS/fix', GPSFix, self.gps)
    	self.imuSub = rospy.Subscriber('/sailboat/IMU', Imu, self.imu)
    	self.windSub = rospy.Subscriber('/sailboat/wind', Pose2D, self.wind)
		self.sailSub = rospy.Subscriber('/sailboat/sail', Float32, self.sail)
    	self.rudderSub = rospy.Subscriber('/sailboat/rudder', Float32, self.rudder)
    	self.rudder2Sub = rospy.Subscriber('/sailboat/rudder2', Float32, self.rudder2)
    	self.velSub = rospy.Subscriber('/sailboat/IMU_Dv', Twist, self.vel)

		if(rospy.get_param("config")):
			utilities.ReadConfig("config/config.txt")
    	time.sleep(1)
    	self.publishMSG("C" + str(mode))
